From d507c4fc3e6e3053e02772b09facbec0eab9e78e Mon Sep 17 00:00:00 2001
From: Build VM <build@example.com>
Date: Mon, 9 Dec 2019 12:57:02 +0000
Subject: [PATCH 01/24] ubunut/makefile

---
 ubuntu/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/ubuntu/Makefile b/ubuntu/Makefile
index 2c1e696f8..3c393e66d 100644
--- a/ubuntu/Makefile
+++ b/ubuntu/Makefile
@@ -16,7 +16,7 @@
 ##
 ##
 ifeq ($(ARCH),x86)
-obj-y                         += vbox/
+obj-n                         += vbox/
 endif
 ##
 ##
@@ -26,7 +26,7 @@ obj-$(CONFIG_HIO)             += hio/
 ##
 ##
 ifeq ($(ARCH),x86)
-obj-y				+= xr-usb-serial/
+obj-n				+= xr-usb-serial/
 endif
 ##
 ##
-- 
2.17.1

