From 2c0166dfee83605cf497eccfc1ac31d9a1c8633b Mon Sep 17 00:00:00 2001
From: Build VM <build@example.com>
Date: Mon, 16 Dec 2019 10:52:06 +0000
Subject: [PATCH 06/24] Wake softirqd, immedeatly...

---
 kernel/softirq.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/kernel/softirq.c b/kernel/softirq.c
index f61113f6b..484654b54 100644
--- a/kernel/softirq.c
+++ b/kernel/softirq.c
@@ -307,9 +307,9 @@ asmlinkage __visible void __softirq_entry __do_softirq(void)
 
 	pending = local_softirq_pending();
 	if (pending) {
-		if (time_before(jiffies, end) && !need_resched() &&
-		    --max_restart)
-			goto restart;
+	//	if (time_before(jiffies, end) && !need_resched() &&
+	//	    --max_restart)
+	//		goto restart;
 
 		wakeup_softirqd();
 	}
-- 
2.17.1

